site/_posts/2023-07-11-get-more-value-out-of-fusionauth.md

```diff
@@ -0,0 +1,321 @@
+
+We can use the FusionAuth APIs to ensure the user is added to the correct group or groups so their permissions will be set properly. After the user is created, we can take this opportunity to automatically assign the user to a specific group or present a front end to the administrator that will guide them in selecting the proper group(s).
+
+<br>
+
+### **Setup**
+Once again we will need to create a few objects to help us organize our groups. The [Add Users to a Group API](/docs/v1/tech/apis/groups#add-users-to-a-group) call we are using requires a collection of groups that each contain an array of user information. In this case, we will only be using the user ID for the new GroupUser information.
+
+```cs
+//For Group object
+public class Group
+{
+    public string ID;
+    public List<GroupUser> Users;
+}
+
+public class GroupUser
+{
+    [JsonProperty(PropertyName = "userId")]
+    public string UserID;
+}
+```
+<br>
+
+### **Input**
+As with creating the user, we need to take the input from the application and assign the values to our Group objects. We can also validate any data we want before actually adding the user to the group. After we are satisfied the user should be added, we can submit the information to the method that will submit our request to the FusionAuth server.
+
+```cs
+private async void btnAddUserToGroup_Click(object sender, EventArgs e)
+{
+    //FusionAuthClient is a class that contains the logic to call  the FusionAuth APIs
+    var faClient = new FusionAuthClient();
+
+    List<GroupUser> users = new List<GroupUser>()
+    {
+        new GroupUser() 
+        {
+            UserID = txtUserIDToAddToGroup.Text
+        }
+    };
+
+    Group group = new Group();
+
+    group.ID = txtGroupIDToAddTo.Text;
+    group.Users = users;
+
+    List<Group> groups = new List<Group>();
+    groups.Add(group);
+
+    //perform custom validation logic here
+    if (IsGroupValid(group))
+    {
+        ReturnValue addUserToGroupReturnValue = await faClient.AddUserToGroup(groups);
+        if (addUserToGroupReturnValue.success == true)
+        {
+            LogResults($"User has been added to the group: {addUserToGroupReturnValue.result}");
+        }
+        else
+        {
+            LogResults($"Adding the user to the group has failed. {addUserToGroupReturnValue.result}");
+        }
+    }
+    else
+    {
+        LogResults($"User and/or group data provided is invalid");
+    }
+}
+```
+
+<br>
+
+### **Add the User to the Group**
+Because the JSON format expected in the request body is not readily serializable given the Group objects we have created, we will have to perform a little wizardry to manipulate the Group objects. While this does not have to do with FusionAuth specifically, I have abstracted the code for that in the FormatGroupUserJSON function. The code for that will be in the project. Once the API call has returned, you can retrieve whichever values you see as relevant and pass it back in the result object.
```

Comment:
Don't tease me! :). you mention `FormatGroupUserJSON` but you don't show that code at all. Please do so.

Reply:
Boy you love code.  I was trying to keep the code focused on the FusionAuth calls, but will add it.

Reply:
Added a Link to the code for those that want to dig deeper.